### ISSUE TITLE
Certain Feeds do not Reliably Automatically Update

### ISSUE THREAD

**Issue by gourdcaptain** [martinrotter/rssguard#148]
#### Brief description of the issue.

Some feeds in RSSGuard like http://kotaku.com/rss do not reliably automatically update - they will sometimes not update for hours unless manually updated.

#### How to reproduce the bug?

1. Add feed to reader. 
2. Set automatic update interval 

#### What is the expected result?

Feed will automatically update when new entries are added to it.

#### What actually happened?

No updates occur (or only partial ones) until the feed is manually updated with right click -> update selected items.

#### Other information (logs, see Wiki)
System: Arch Linux 64-bit, RSSGuard 3.5.0.

Logs are currently being looked for - this issue due to being irregularly occurring is hard to reproduce on command. Will post logs once acquired.

**Comment by gourdcaptain** [comment]
[rssguard-log.txt](https://github.com/martinrotter/rssguard/files/1377634/rssguard-log.txt)
Here's a log. Incident occured right before I closed it - I force updated the Kotaku feed and about two hours of articles missed by 30 minute interval updates showed up.

**Comment by martinrotter** [comment]
Does it happen with particular feeds only, or is this more general problem?

**Comment by gourdcaptain** [comment]
It happens with only certain feeds. That one (Kotaku) is the most drastic case by far, though.

**Comment by martinrotter** [comment]
Okay, that is interesting.

**Comment by martinrotter** [comment]
OK, I added the feed and set auto-update global per 5 minutes, let's see what happens.

**Comment by martinrotter** [comment]
First auto-run, obtained 24 messages.

What happens when you do not get messages? Does feeds'name in feeds list turn into red color? Is it possible that the feed has big ping? If so, try to increase max timeout for downloading - Settings - Feeds/messages - Timeout...., change it do XXXX miliseconds, for example 45000

**Comment by gourdcaptain** [comment]
When it has the error, it doesn't turn red. It just silently fails to add new messages until it's manually updated.

**Comment by martinrotter** [comment]
OK, second run - 0 new messages. Will check your log.

**Comment by gourdcaptain** [comment]
Unfortunately, we're kind of at the mercy of the feed being updated with new posts by the site to test this...

On October 12, 2017 1:54:28 AM EDT, martinrotter <notifications@github.com> wrote:
>OK, second run - 0 new messages. Will check your log.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_martinrotter_rssguard_issues_148-23issuecomment-2D336028257&d=DwICaQ&c=3buyMx9JlH1z22L_G5pM28wz_Ru6WjhVHwo-vpeS0Gk&r=0zQ8RN820LSg0alBA3zG9VWeCI6KR1FUJkKAfmvCcAs&m=7UikWgP4_UKVeYPx6vgKaftM1Zx-iei-MVSN_HoDidI&s=5g5byDvL8jfB-1pUXH64GRBWYzZzG43vEGdjv37zVQw&e=
>

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by martinrotter** [comment]
Yes, I will have to debug this on Linux when I get home, my Windows debugger is somehow buggy. Initially, it seems to work fine, logs do not suggest anything wrong.

Can you maybe try with only single feed (the one) added in your RSS Guard? So that log is smaller and not so cluttered?

**Comment by gourdcaptain** [comment]
Is there a way to run rssguard from a different config temporarily so I don't have to mess with my (hilariously full) main configuration I'm using all the time? I can do a backup/restore if I have to, though.

**Comment by martinrotter** [comment]
No, I would recommend to check in about dialog, what user folder is used, and then quit rssguard, rename your folder (make backup) and start rssguard with clean folder.

**Comment by martinrotter** [comment]
The logic which updates feeds is quite complicated (feed's data are downloaded simultaneously for higher speed, work with threading etc.) but right now, I just do not see the problem. I will wait for your simplified log with only that one particular feed added and in meanwhile I will try here.

**Comment by gourdcaptain** [comment]
[rssguard-log.txt](https://github.com/martinrotter/rssguard/files/1383662/rssguard-log.txt)
Here's the minimal log you asked for that managed to replicate the bug over about an hour and a half after I set it up. I tried to replicate the settings of the profile that had the bug on my main system (SQLite DB, in memory) and it worked with the one Kotaku feed only in the profile.

**Comment by martinrotter** [comment]
@gourdcaptain SO the issue is still present even when only single feed is added? Not sure if I understood your point correctly.

**Comment by gourdcaptain** [comment]
Yes, the issue occurs with only the one feed.

**Comment by martinrotter** [comment]
@gourdcaptain Just to mention, I managed AppImages generation to work.

https://github.com/martinrotter/rssguard/wiki/Linux-development-builds

I will investigate the log and let you know my detailed findings tomorrow.

**Comment by gourdcaptain** [comment]
Links there don't work - they've got an extra newline character "\n" in the URL. AppImages aren't really my thing, in part because there's an issue with QT5 and them for me I'm not sure if there's a good workaround/solution for. I use the qt5-styleplugins package on Arch to use the GTK2 theme for QT5 to make it match the theme of my XFCE environment (it's got a GTK3 version, but there isn't a GTK3 QT5 theme), and the AppImages don't include that theme, so AppImage'd QT5 apps turn out with a very ugly and poorly scaled for my monitor default theme.

**Comment by martinrotter** [comment]
Always try links on the bottom, they are newest.

**Comment by martinrotter** [comment]
Yes, AppImages are not very common thing on Arch either. I do not use them too, AUR is just much more native for Arch and superior.

I will release new RSS Guard with some bug fixed today.

So, I checked your log, line by line:
1. You use SQLite in-memory working DB (no problem with that).
2. You have active auto-update each 30 minutes.
3. On the first run, 22 of messages for feed Kotaku was downloaded, 2 of them were added to database, 20 of them were probably already existing in the DB.
4. On the second auto-update run, 22 messages were downloaded again, 1 of them was probably new and was added to RSS Guard database, other messages probably were not new.
5. On the third auto-update run, 22 messages were downloaded, 2 of them were added to database, other messages were not new. 
6. You quit the app.

It seems like really standard application workflow, no errors in log. Also when I tried for myself to manually update feed like each 10 minutes, I also obtained like 0-2 new messages, so it seems that they add/change like 1 message in Kotaku feed per 30 minutes.

Were your observations the same as my findings from the log? Did you really see new two messages after first auto-update wave and 1 new in second wave and 2 new messages in third wave? It seems that Kotaku feed file is correctly downloaded and parsed each time.

So what exactly is the issue? You have impression that when auto-update is ran, not all messages are downloaded? Does this happen with other feeds? Does this happen more if you have many feeds in RSS Guard? Do you feel it is related to database size?

**Comment by gourdcaptain** [comment]
Yeah while there are new messages, the updates are not all grabbed automatically - some of the messages are backdated on a manual update prior to the last automatic update, meaning they were missed by the automatic update. They are all fetched in another RSS reader on my phone (Press) if it updates at the same time as the automatic update. It does occur with other feeds like this one, but rarer: https://www.techdirt.com/techdirt_rss.xml

**Comment by martinrotter** [comment]
So, if I keep auto-updates running for like two hours (and I get like 5 new messages) and then I will update feed manually, I should get many new messages (if the bug is present), yes?

**Comment by gourdcaptain** [comment]
You'll get a few? Usually? It's not 100% reliable, but that's the way it behaves.

**Comment by martinrotter** [comment]
I will perform testing with Inoreader, to get detailed information about that feed. Currently it says that LAST post to feed Kotaku was made 4 hours ago. To me, it really does not seem that Kotaku feed is getting major numbers of new messages often, but we will see, I will let it run for a few hours.

**Comment by martinrotter** [comment]
So, I let RSS Guard run for three hours and simultaneously I ran Inoreader. Both programs gathered same messages from your feed.
![screenshot 2017-10-16 12 48 20](https://user-images.githubusercontent.com/1255302/31608362-6d466f08-b270-11e7-8a6f-8dbe821ee083.png)

Honestly, I do not know what to look for. Screenshot says that RSS Guard works, your latest log also says that RSS Guard works.

**Comment by martinrotter** [comment]
I can reliably say, that Kotaku feed is not getting any big amount of new messages in short time intervals, it seems to get max 1-2 new messages per hour, probably less and that is the situation I observed in both independet applications (RSS Guard, inoreader).

**Comment by gourdcaptain** [comment]
Anything I can do? Before and after database backups?

**Comment by martinrotter** [comment]
So, when you recorded the last log, how many new messages did you see in Kotaku feed in RSS Guard during that testing period (1 hour?)?

**Comment by gourdcaptain** [comment]
I don't remember beyond "there were a couple when I checked, and a few more filled in that were missed when I did the manual update."

On October 16, 2017 9:02:17 AM EDT, martinrotter <notifications@github.com> wrote:
>So, when you recorded the last log, how many new messages did you see
>in Kotaku feed in RSS Guard during that testing period (1 hour?)?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_martinrotter_rssguard_issues_148-23issuecomment-2D336879679&d=DwICaQ&c=3buyMx9JlH1z22L_G5pM28wz_Ru6WjhVHwo-vpeS0Gk&r=0zQ8RN820LSg0alBA3zG9VWeCI6KR1FUJkKAfmvCcAs&m=at6FbQPKJ90LOejrBUJxofg5ghVQTeLeKdBqttNUODw&s=5rSlPY40_wqWxiDOXfLnzjzlOOnrxOywKgVcopVqtYQ&e=
>

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by martinrotter** [comment]
I believe, the last part when you at the end manually updated the feed, is not in the log file you uploaded, it would be nice to have complete log from complete workflow:
1. You have rssguard running with one feed for several hours.
2. Your expected behavior is now visible (Kotaku feed auto-gets only some new messages).
3. You now after few hours manually update.
4. You see big number of new messages.
5. You send log.

I believe that point 3 is not recorded in log, maybe you stopped too soon. Anyway, as I say, there is really nothing to solve because the log looks crisp clean. In every auto-update run (30 minutes) 1-2 new messages are downloaded. It would be completely different story if I saw that connection to feed file fails or messages are not inserted into DB, but no such thing happened according to uploaded log.

You can of course send DB backups before/after and I will check them, but I am not sure what I can find.

**Comment by gourdcaptain** [comment]
Well, I did do the number three bit before closing RSSGuard... But I can tell I'm either bad at communicating something or it's some weird particularity across both machines I have this issue on at this point. :(

On October 16, 2017 9:17:40 AM EDT, martinrotter <notifications@github.com> wrote:
>I believe, the last part when you at the end manually updated the feed,
>is not in the log file you uploaded, it would be nice to have complete
>log from complete workflow:
>1. You have rssguard running with one feed for several hours.
>2. Your expected behavior is now visible (Kotaku feed auto-gets only
>some new messages).
>3. You now after few hours manually update.
>4. You see big number of new messages.
>5. You send log.
>
>I believe that point 3 is not recorded in log, maybe you stopped too
>soon. Anyway, as I say, there is really nothing to solve because the
>log looks crisp clean. In every auto-update run (30 minutes) 1-2 new
>messages are downloaded. It would be completely different story if I
>saw that connection to feed file fails or messages are not inserted
>into DB, but no such thing happened according to uploaded log.
>
>You can of course send DB backups before/after and I will check them,
>but I am not sure what I can find.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_martinrotter_rssguard_issues_148-23issuecomment-2D336883557&d=DwICaQ&c=3buyMx9JlH1z22L_G5pM28wz_Ru6WjhVHwo-vpeS0Gk&r=0zQ8RN820LSg0alBA3zG9VWeCI6KR1FUJkKAfmvCcAs&m=oparl8MdOQWr2yJCbhH7At8r7kH8a997retjo5dvV5c&s=xx2jYz8zV4r2KERv_78XnD8gR8jgsjIYV-ZpJaAfjdE&e=
>

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by martinrotter** [comment]
Yes, I missed it. You manually update just before end of log file. During that update 22 messages were downloaded and 2 of them were new and were added to database, so from my point of view (from log perspective) everything works.

**Comment by martinrotter** [comment]
I made some adjustments to feed update process. Can you please:
1. Download latest dev build.
2. Start it with clean user data.
3. Add feeds (kotaku and some others?).
4. Configure auto-updates.
5. Reproduce the bug.
6. Send me log.

THANK YOU. :)

**Comment by gourdcaptain** [comment]
[rssguard-log.txt](https://github.com/martinrotter/rssguard/files/1431329/rssguard-log.txt)
Here you go. Didn't get around to this for a few days, and the log's a bit long because I left it going overnight even with just one feed in the setup. Hope it helps. :)

**Comment by danowar2k** [comment]
I've got a similar issue and filed it. Maybe I will be able to pinpoint it...

**Comment by danowar2k** [comment]
Reading the log, it would probably be helpful if the debugging lines wouldn't only contain the ID of feeds but also the URLs. Then people could point at lines where they think the error happens...

**Comment by martinrotter** [comment]
"would probably be helpful if the debugging lines wouldn't only contain the ID of feeds but also the URLs"

Which lines exactly could be that, send those lines, I will add URLs to them.

**Comment by danowar2k** [comment]
Let's see...going by the log above, it would probably be good if...
- The log contained the RSSguard version number so that you would know it

It would be good if the following lines contained the URL of the feed, or rather more info to identify what is meant:
- [rssguard] DEBUG: Loading messages from feeds: '1'. (17/31/10 01:03:27)
- [rssguard] DEBUG: Downloading new messages for feed "1" in thread: '0x7fd1cbfff700'. (17/31/10 01:03:29)
- [rssguard] DEBUG: Downloaded 23 messages for feed "1" in thread: '0x7fd1cbfff700'. (17/31/10 01:03:29)
- [rssguard] DEBUG: Saving messages of feed "1" in thread: '0x7fd1de96f700'. (17/31/10 01:03:29)
- [rssguard] DEBUG: 23 messages for feed 1 stored in DB. (17/31/10 01:03:29)
- [rssguard] DEBUG: Made progress in feed updates, total feeds count 1/1 (id of feed is 1). (17/31/10 01:18:09)

Here a list of feeds would be good maybe:
- [rssguard] DEBUG: There is request to reload feed model, reloading the 1 items individually. (17/31/10 01:03:50)

Or instead of the feed URL, the feed title perhaps...

Which files contain the logic for updating a feed? Maybe more eyes than two will help?

**Comment by danowar2k** [comment]
Ah, found the method, it seems....
https://github.com/martinrotter/rssguard/blob/master/src/miscellaneous/feedreader.cpp

**Comment by danowar2k** [comment]
Perhaps additional debug messages...
if (m_feedDownloader == nullptr) {
    qDebug("Initializing FeedDownloader...");
    m_feedDownloader = new FeedDownloader();

or here:
https://github.com/martinrotter/rssguard/blob/master/src/core/feeddownloader.cpp
in the "updateAvailableFeeds" method...
 while (!m_feeds.isEmpty()) {
    connect(m_feeds.first(), &Feed::messagesObtained, this, &FeedDownloader::oneFeedUpdateFinished,
            (Qt::ConnectionType)(Qt::UniqueConnection | Qt::AutoConnection));

    if (m_threadPool->tryStart(m_feeds.first())) {
      m_feeds.removeFirst();
      m_feedsUpdating++;
    }
    else {
      // We want to start update of some feeds but all working threads are occupied.
      break;
    }
  }

If you can't update the feeds because all threads are occupied you wait until you can. So that doesn't seem to be a problem because the app would just hang after some time if it were....You could add a timeout for updating the feeds after which you log a message that you couldn't update the feeds...

Where is the method "connect" defined?

**Comment by danowar2k** [comment]
Next up:
https://github.com/martinrotter/rssguard/blob/master/src/services/abstract/feed.cpp

Let's see...Feeds update when run. As you have multiple feed classes inheriting from Feed, you could add additional debug information about a specific feed in the update message (e.g. feed type, standard or Gmail or whatever)

```
else if (status() != NewMessages) {
    *error_during_obtaining = false;
  }
```
I see that the enum for the status can be normal, NewMessages and various errors. Here it seems they are ignored?

**Comment by danowar2k** [comment]
I just checked, after returning from energy saving I can't seem to update my Spiegel feed.

Also, you add debug messages containing URLs to article thumbnails. It would help for debugging to have the title of the article for reference. At the moment I can't see which 30 articles are downloaded by the feed.
```
[rssguard] DEBUG: Starting feed updates from worker in thread: '0x2760'. (17/09/11 22:31:30)
[rssguard] DEBUG: Downloading new messages for feed "48" in thread: '0xd5c'. (17/09/11 22:31:30)
[rssguard] DEBUG: Settings of BaseNetworkAccessManager loaded. (17/09/11 22:31:30)
[rssguard] DEBUG: Destroying SilentNetworkAccessManager instance. (17/09/11 22:31:30)
[rssguard] DEBUG: Adding enclosure 'http://www.spiegel.de/images/image-1211932-thumbsmall-fzde.jpg' for the message. (17/09/11 22:31:30)
[...]
[rssguard] DEBUG: Adding enclosure 'http://www.spiegel.de/images/image-1211610-thumbsmall-fdkg.jpg' for the message. (17/09/11 22:31:30)
[rssguard] DEBUG: Downloaded 30 messages for feed "48" in thread: '0xd5c'. (17/09/11 22:31:30)
[rssguard] DEBUG: Saving messages of feed "48" in thread: '0x2760'. (17/09/11 22:31:30)
[rssguard] DEBUG: Updating messages in DB. Main thread: 'false'. (17/09/11 22:31:30)
[rssguard] DEBUG: 0 messages for feed 48 stored in DB. (17/09/11 22:31:30)

```
By reading the rss feed in a browser, I know I am missing one new article. But I don't know if the article was downloaded or missed.

**Comment by martinrotter** [comment]
> If you can't update the feeds because all threads are occupied you wait until you can.

Method to update rest of feeds in the queue is re-triggered when some feed's update finishes see method:

`void FeedDownloader::oneFeedUpdateFinished(const QList<Message>& messages, bool error_during_obtaining)`

**Comment by martinrotter** [comment]
Virtual method `Feed::obtainNewMessages(bool* error_during_obtaining)` always downloads all messages from feed source XML online file.

The lines
```
  else if (status() != NewMessages) {
    *error_during_obtaining = false;
  }
```
are there to ensure that if feed already contains new messages (and has blue color in the list of feeds) then make sure that you do not set the feed to error state (color would turn to red) ---> I removed this code now, it is confusing a bit - if there is error during feed update, feed will now turn red in all cases (but will be bold red if there are any unread messages).

**Comment by martinrotter** [comment]
> lso, you add debug messages containing URLs to article thumbnails

Not currently possible because the logic which parses individual XML chunks and retrieves attachments from it is completely separate from message logic, but you can easily now which feed it belongs because in this inserting of messages to DB is synchronous. SO if you see this in the log:

```
[rssguard] Adding enclosure 'http://feedproxy.google.com/~r/InterestingThingOfTheDay/~5/R8q5WACGZPM/ITotD-334-Array.mp3' for the message. (17/10/11 07:14:34)
  Type: DEBUG
  File: ..\rssguard\src\services\standard\rssparser.cpp (line 61)
  Function: class Message __cdecl RssParser::extractMessage(const class QDomElement &,class QDateTime) const

[rssguard] Adding enclosure 'http://feedproxy.google.com/~r/InterestingThingOfTheDay/~5/h_imBKPukfQ/ITotD-333-Audium.mp3' for the message. (17/10/11 07:14:34)
  Type: DEBUG
  File: ..\rssguard\src\services\standard\rssparser.cpp (line 61)
  Function: class Message __cdecl RssParser::extractMessage(const class QDomElement &,class QDateTime) const

[rssguard] Adding enclosure 'http://feedproxy.google.com/~r/InterestingThingOfTheDay/~5/3ZG1NwRSirM/ITotD-332-Flywheel.mp3' for the message. (17/10/11 07:14:34)
  Type: DEBUG
  File: ..\rssguard\src\services\standard\rssparser.cpp (line 61)
  Function: class Message __cdecl RssParser::extractMessage(const class QDomElement &,class QDateTime) const

[rssguard] Downloaded 15 messages for feed ID "4" URL: "http://feeds.feedburner.com/InterestingThingOfTheDay" title: "Interesting Thing of the Day" in thread: '0x54'. (17/10/11 07:14:34)
  Type: DEBUG
  File: ..\rssguard\src\services\abstract\feed.cpp (line 174)
  Function: void __cdecl Feed::run(void)
[rssguard] Adding enclosure 'http://feedproxy.google.com/~r/InterestingThingOfTheDay/~5/R8q5WACGZPM/ITotD-334-Array.mp3' for the message. (17/10/11 07:14:34)
  Type: DEBUG
  File: ..\rssguard\src\services\standard\rssparser.cpp (line 61)
  Function: class Message __cdecl RssParser::extractMessage(const class QDomElement &,class QDateTime) const

[rssguard] Adding enclosure 'http://feedproxy.google.com/~r/InterestingThingOfTheDay/~5/h_imBKPukfQ/ITotD-333-Audium.mp3' for the message. (17/10/11 07:14:34)
  Type: DEBUG
  File: ..\rssguard\src\services\standard\rssparser.cpp (line 61)
  Function: class Message __cdecl RssParser::extractMessage(const class QDomElement &,class QDateTime) const

[rssguard] Adding enclosure 'http://feedproxy.google.com/~r/InterestingThingOfTheDay/~5/3ZG1NwRSirM/ITotD-332-Flywheel.mp3' for the message. (17/10/11 07:14:34)
  Type: DEBUG
  File: ..\rssguard\src\services\standard\rssparser.cpp (line 61)
  Function: class Message __cdecl RssParser::extractMessage(const class QDomElement &,class QDateTime) const

[rssguard] Downloaded 15 messages for feed ID "4" URL: "http://feeds.feedburner.com/InterestingThingOfTheDay" title: "Interesting Thing of the Day" in thread: '0x54'. (17/10/11 07:14:34)
  Type: DEBUG
  File: ..\rssguard\src\services\abstract\feed.cpp (line 174)
  Function: void __cdecl Feed::run(void)
```

Then you know that these enclosures were obtained for feed ID "4". I also fixed now the string "Adding enclosure..." which actually means that the enclosure was decoded/obtained from feed XML (it does not mean it was added to DB). When any message is added/updated in DB, all its enclosures are automatically overwritten too.

So, I commited it.

Please test it when the development build gets compiled - see wiki/development builds. :) THANK YOU.

**Comment by martinrotter** [comment]
Btw, according to your log it seems that after wake from suspend, messages were actually downloaded

[rssguard] DEBUG: Adding enclosure 'http://www.spiegel.de/images/image-1211610-thumbsmall-fdkg.jpg' for the message. (17/09/11 22:31:30)
[rssguard] DEBUG: Downloaded 30 messages for feed "48" in thread: '0xd5c'. (17/09/11 22:31:30)
[rssguard] DEBUG: Saving messages of feed "48" in thread: '0x2760'. (17/09/11 22:31:30)
[rssguard] DEBUG: Updating messages in DB. Main thread: 'false'. (17/09/11 22:31:30)
[rssguard] DEBUG: 0 messages for feed 48 stored in DB. (17/09/11 22:31:30)

but not updated in DB (same message probably already existed in DB).

OR Maybe only some little detail about message has changed (currently, messages are uniquely identified with CONTENTS and CREATION DATE) so if any of these attributes is updated in online XML, it should also update itself in RSS Guard. If for example title or author is different, then RSS Guard treats the message as completely new message.

**Comment by martinrotter** [comment]
One more question from: Do you feel that big number of messages is not updated when the issue occurs? Or are we talking about some small number of individual messages? I am asking to determine in which subsystem the issue might be (if it is in actual feed XML file downloading/parsing or in saving of individual new/updated messages into DB).

So try to reproduce with latest version which has more debugging output in those sections of code you proposed and some more.

https://github.com/martinrotter/rssguard/commit/a93c14ddc51bb7ffbbf4e25b31808c985b05fc2d

**Comment by danowar2k** [comment]
First: It seems we may have a language problem: English isn't my first language (but German is). I can't speak Dutch (I thought I saw somewhere you are from the Netherlands), but we could try German if you'd like. If not, doesn't matter, I'll try to write more careful from now on...

**Comment by martinrotter** [comment]
I am from Czech Republic. :).

Why do you think we have language problem?

**Comment by danowar2k** [comment]
>> If you can't update the feeds because all threads are occupied you wait until you can.

> Method to update rest of feeds in the queue is re-triggered when some feed's update finishes see method:

I'll look into that method, but I seem to have phrased it wrong. What I meant:
Looking at your code, if one feed cannot update because the feed can get no thread, the feed will wait until it can secure a thread for itself.

Ah, so `obtainNewMessages` always downloads all articles contained in the feed and doesn't just parse each item and check if download is necessary (don't know if each method would cause much overhead, though).

Quick idea: If feeds turn red, an error happened during update. Could you add a context menu item to a red feed that shows the last error message? 

>> Why do you think we have language problem?

> but you can easily now which feed it belongs because in this inserting of messages to DB is synchronous

Because I didn't understand this, for example. It seems sometimes there are words missing? What is meant by "this" and "it"?

Also, I'm not careful enough when I write comments. Programming is precise and I seem to overlook stuff like writing "feeds" instead of "feed".

>> Also, you add debug messages containing URLs to article thumbnails. It would help for debugging to have the title of the article for reference. At the moment I can't see which 30 articles are downloaded by the feed.

>Not currently possible because the logic which parses individual XML chunks and retrieves attachments from it is completely separate from message logic

When you display the message
```
[rssguard] Downloaded 15 messages for feed ID "4" URL: "http://feeds.feedburner.com/InterestingThingOfTheDay" title: "Interesting Thing of the Day" in thread: '0x54'. (17/10/11 07:14:34)
```
you know the number of downloaded messages. Can't you iterate over them and display the titles at that moment?

**Comment by danowar2k** [comment]
I'll try the new build now...

**Comment by danowar2k** [comment]
Before I do, after obtaining new messages, you iterate over the messages and extract title etc.
You could output the stuff there for debugging.

**Comment by martinrotter** [comment]
> Looking at your code, if one feed cannot update because the feed can get no thread, the feed will wait until it can secure a thread for itself.

It in fact does wait - but "spinning" approach is used.

> don't know if any of the two methods would cause more overhead, though

That's what I thought too. It is probably best to download all messages from channel (you have to download whole XML anyway).

> Quick idea: If feeds turn red, an error happened during update. Could you add a context menu item to a red feed that shows the last error message?
When you hover your mouse over the feed, you should see some error-related info in tooltip.

> Because I didn't understand this, for example. It seems sometimes there are words missing? What is meant by "this" and "it"?
Yes,  I fucked it up. Le me rephrase.

"""
Not currently possible because the logick, which parses individual XML chunks and decodes attachments from it is completely separate from message logic, but you can easily SEE which FEED it the above log lines belongs TO, because parsing of messages is synchronous. So if you see that "Added enclosure..." log entry then search for next "Downloaded XX messages for feed..." log entry and you see which feed enclosures are for.
"""

> Also, I'm not careful enough when I write comments.
Me neither.

> you know the number of downloaded messages. Can't you iterate over them and display the titles at that moment?

I could, but I added extra debug log entries so that now you will actually see message titles when it matters - when message is added or updated. But if we fail to identify the issue correctly, I will add those debug entries too.

**Comment by danowar2k** [comment]
Okay, I'm trying the debug build with -log.

Something is strange with the logging. When I start the program, the rssguard.log is there, but Windows 7 explorer says it has 0 bytes. When I open the log in Notepad++, it is empty. Only after waiting a few more seconds and doing F5 to update the explorer the file gains size and I can reload it in Notepad++.

Also strange: I marked a single feed (having 52 fresh unread messages), told RSSguard to "update the selected feeds", waited a bit and reopened the log. The same 52 messages were downloaded, but at that moment only 21 database checks were logged. Only when I updated the feed again (!), the other 31 log messages were written to the file.

Are you doing some delayed flushing when appending to the log? Can't you force flush after each message when the program is started with "-log"?

> I could, but I added extra debug log entries so that now you will actually see message titles when it matters - when message is added or updated. But if we fail to identify the issue correctly, I will add those debug entries too.

I hope your debug messages will be enough for now, they have the necessary info to identify downloaded articles, it seems.

Now I'll wait when the error comes up again...

**Comment by danowar2k** [comment]
We could use some IM communication if you'd prefer.

**Comment by danowar2k** [comment]
"L?stige Marotten: Was nervt Sie am Handy am allerallermeisten?"

Noticed this. I hope you use UTF-8 everywhere....

**Comment by danowar2k** [comment]
Instead or waiting for shit to happen in the real world ;-) you could provide a test feed of sorts that you build. I could then test with that...

**Comment by martinrotter** [comment]
I comitted fix for those unflushed logs.

On, Windows logging to file is sometimes crazy nightmare, but now it is explicitly flushed after each write. Should work better now.

As for encoding. I am using older qDebug("string") API, instead of new qDebug() << "str" API. Older API is more C-like and bit faster and I am used to it so far. Problem of OLD api is that it has some problems with special characters but it is really not important for logging, ASCII is enough.

Let me see your findings.

**Comment by danowar2k** [comment]
Downloading the new build...

**Comment by danowar2k** [comment]
This happened: I had 30 messages in the feed, I updated the feed, then I had 29 messages in the feed.
Log:
````
[rssguard] DEBUG: Checking if message with title 'Katalonien: Ex-Parlamentspr?sidentin kann U-Haft gegen Kaution vermeiden', url 'http://www.spiegel.de/politik/ausland/katalonien-carme-forcadell-droht-u-haft-a-1177294.html#ref=rss' and author '' is present in DB. (17/10/11 10:48:48)
[rssguard] DEBUG: Message with these attributes is already present in DB and has DB ID 1426547112. (17/10/11 10:48:48)
[rssguard] DEBUG: Updating message with title 'Katalonien: Ex-Parlamentspr?sidentin kann U-Haft gegen Kaution vermeiden' url 'http://www.spiegel.de/politik/ausland/katalonien-carme-forcadell-droht-u-haft-a-1177294.html#ref=rss' in DB. (17/10/11 10:48:48)
[rssguard] DEBUG: 1 messages for feed 11 stored in DB. (17/10/11 10:48:48)
[...]
[rssguard] DEBUG: Made progress in feed updates, total feeds count 1/1 (id of feed is 11). (17/10/11 10:48:48)
````
Then after another update a few minutes later it reappeared...

````
[rssguard] DEBUG: Checking if message with title 'Katalonien: Ex-Parlamentspr?sidentin kann U-Haft gegen Kaution vermeiden', url 'http://www.spiegel.de/politik/ausland/katalonien-carme-forcadell-droht-u-haft-a-1177294.html#ref=rss' and author '' is present in DB. (17/10/11 10:51:47)
[rssguard] DEBUG: Adding new message with title 'Katalonien: Ex-Parlamentspr?sidentin kann U-Haft gegen Kaution vermeiden' url 'http://www.spiegel.de/politik/ausland/katalonien-carme-forcadell-droht-u-haft-a-1177294.html#ref=rss' to DB. (17/10/11 10:51:47)
[...]
[rssguard] DEBUG: SQLite connection 'feed_upd' is already active. (17/10/11 10:51:47)
[rssguard] DEBUG: File-based SQLite database connection 'feed_upd' to file 'E:\Eigenes\Desktop\rssguard-3.5.5-eb777b9-nowebengine-win64\data\database\local\database.db' seems to be established. (17/10/11 10:51:47)
[rssguard] DEBUG: 1 messages for feed 11 stored in DB. (17/10/11 10:51:47)
[rssguard] DEBUG: Made progress in feed updates, total feeds count 1/1 (id of feed is 11). (17/10/11 10:51:47)
````
But I think it may just be that they took the article from the feed because something had to be corrected. News...

**Comment by martinrotter** [comment]
Look at list of messages, do you see that message twice? Does it have same date/title/author? Do they differ in any way?

**Comment by martinrotter** [comment]
> I really think some testing area would help where you generate a RSS feed which has a new item every minute. Their titles would show the generated time.
> 
> Then tests could be run against that RSS feed.
> 
> Maybe have a few more RSS feeds which test other things (randomly update/edit the same article, have non-ASCII chars somewhere, violate some RSS standard, etc etc)

This would certainly be nice, but I am quite busy, someone would have to write some webservice which would server these testing RSS feeds.

**Comment by danowar2k** [comment]
Dammit, it doesn't seem to happen when the PC is only a few minutes in energy saving mode. Or maybe activating the log prevents it from happening.
Or maybe it only happens when the database has a certain size.
Or maybe it only happens when RSSguard has been in memory for a longer time.

Can I use the database from my normal version for debugging? Which files do I have to copy?

**Comment by danowar2k** [comment]
> Look at list of messages, do you see that message twice? Does it have same date/title/author? Do they differ in any way?
No the old messages got removed from the article list. Author was none both times, though, I think.
I can't compare the two messages because the old one is gone :-(

**Comment by martinrotter** [comment]
You can copy "database.db" (see About dialog), it should work. Can you send me URL of the feed from the log above?

**Comment by martinrotter** [comment]
Btw, IM channel for rssguard is IRC channel #rssguard on freenode IRC. I am there when I am at home (usually in evenings when my child is sleeping - I do not runy my home PC workstation when he is not sleeping). :)

**Comment by danowar2k** [comment]
> Can you send me URL of the feed from the log above?

http://www.spiegel.de/schlagzeilen/tops/index.rss

**Comment by danowar2k** [comment]
Ok, so I tried forcing the bug with my normal RSSguard 3.5.4 without a log. I went into energy save mode, came back and the feed still updated properly.

Now I have had the program running for a half hour or so. The feed has three new articles but when I manually update it, they aren't added to the article list. As no log is running, I can't see what's happening, of course.

I will now restart the dev build with my usual database and do the same (go into energy save, come back, wait for an hour and try to update).

If this doesn't reproduce the bug, I bet there is something happening after that time or after the energy save that doesn't happen when you activate logging.

I would bet on file handles that are closed and never reopened, or threads dying and not replaced or network connections closed but never checked to reopen them.

Somewhere in the code happens something to stuff in memory where you don't check if things are ready to use, or you have written such fail-safes that the program seems to work but only because you continue without throwing any errors when things aren't there that should be there.

If any uncatched error would be there, the program would produce crashes or a backtrace somewhere, wouldn't it?

**Comment by danowar2k** [comment]
At the moment it seems it would really help if I could start/stop logging to a file while the program is running...

**Comment by martinrotter** [comment]
Yes, any uncaught exceptions lead to a crash.

Any network connections for downloading feed XMLs are created and closed only when really needed. Not sure about feed downloader thread, it really could get somehow corrupted when PC wakes up from sleep. I will have to try myself.

1. You can launch rssguard in command line (without -log option) and redirect its output into file - `rssguard.exe > log.txt 2>&1`.
2. Put PC to sleep with RSS Guard and command line running.
3. Restore from sleep.
4. Reproduce bug and stop RSS Guard.
5. Upload log.txt.

**Comment by danowar2k** [comment]
Ok, so using the dev build without "-log", going to energy save, returning from it and then waiting for half an hour or so leads to the same result: the feed isn't updated anymore, even when manually updating.

I'll now try with both methods above, the "-log" method and the redirect stdout and stderr method, using dev build + my database...

**Comment by martinrotter** [comment]
OK, once you will be sure you have precise steps to reliably reproduce the bug, paste them here and I will try to exactly follow them.

**Comment by danowar2k** [comment]
I've reproduced the bug with my 3.5.4, my database. I attach the log. But 3.5.4 has not enough debug messages, so I can't see anything wrong there, except I'm wondering this:
If the Spiegel feed does have 30 articles in it, why are there only ca. 21 thumbnail lines like this:

```
[rssguard] DEBUG: Adding enclosure 'http://www.spiegel.de/images/image-1212287-thumbsmall-lwpk.jpg' for the message. (17/10/11 14:00:42)
```

Attached: [rssguard.log](https://github.com/martinrotter/rssguard/files/1461385/rssguard.log)

Now, back to the dev build, using the stderr redirect method.

**Comment by martinrotter** [comment]
Maybe not all messages had enclosures.

**Comment by danowar2k** [comment]
Okay, I've changed for testing to this feed here:
http://www.spiegel.de/schlagzeilen/index.rss
This has ALL headlines from Spiegel.de , so it is updated more frequently. The bug still happened with this feed in 3.5.4, so I'm back at the dev build, trying to force it.

**Comment by danowar2k** [comment]
I will also test if I even need going to Energy Save mode in 3.5.4 for the bug to manifest.

If the bug isn't reproducible in the dev build, then any code between 3.5.4 and the dev build has either had a fix for the bug or all those additional debug messages keep something alive that otherwise breaks.

**Comment by danowar2k** [comment]
Ok at least I can say that going into Energy Save mode isn't necessary to produce the bug in 3.5.4 .

Now back to the dev build (even if I don't know what I should do anymore now)

**Comment by martinrotter** [comment]
Shit, i missed this comment, SORRY. Let me respond now. :)

I did a couple of code changes regarding overall RSS Guard behavior which could make it better regarding this bug report and it is possible that the bug was really partially fixed.

Feel free to send any logs/databases I will check them and maybe it could lead somewhere.

So, I will not close this yet, but just asking: do you thing (try latest 3.5.5 version) the bug is actually fixed?

**Comment by danowar2k** [comment]
I will try it out and report.

2017-11-30 10:17 GMT+01:00 martinrotter <notifications@github.com>:

**Comment by gourdcaptain** [comment]
Seems to be working for me at the moment? No egregious cases of the bug have come up since 3.5.5

**Comment by martinrotter** [comment]
Is this reliably working presently?

**Comment by gourdcaptain** [comment]
Yep. Although I've changed my setup significantly since I originally reported this - new desktop, went from btrfs on a hard drive as the filesystem the database was on to ext4 on a SSD, but it was working reliably before the switch I think.

**Comment by martinrotter** [comment]
I create merged ticket for similar problems which have probably the same root cause - head over there to discuss or inform me of development on your side.

https://github.com/martinrotter/rssguard/issues/237
